lib/IRGen/GenPack.cpp

```diff
@@ -359,15 +358,222 @@ irgen::emitTypeMetadataPackRef(IRGenFunction &IGF, CanPackType packType,
+  auto instantiatedPatternTy =
+      elementEnv
+          ->mapPackTypeIntoElementContext(patternTy->mapTypeOutOfContext())
+          ->getCanonicalType();
+
+  // Emit the element witness table.
+  auto *wtable = emitWitnessTableRef(IGF, instantiatedPatternTy,
+                                     /*srcMetadataCache=*/nullptr, conformance);
+  return wtable;
+}
+
+static void emitExpansionWitnessTablePack(IRGenFunction &IGF, Address pack,
+                                          CanPackExpansionType expansionTy,
+                                          ProtocolConformanceRef conformance,
+                                          llvm::Value *dynamicIndex,
+                                          llvm::Value *dynamicLength) {
+  auto *prev = IGF.Builder.GetInsertBlock();
+  auto *check = IGF.createBasicBlock("pack-expansion-check");
+  auto *loop = IGF.createBasicBlock("pack-expansion-loop");
+  auto *rest = IGF.createBasicBlock("pack-expansion-rest");
+
+  IGF.Builder.CreateBr(check);
+  IGF.Builder.emitBlock(check);
+
+  // An index into the source witness table pack.
+  auto *phi = IGF.Builder.CreatePHI(IGF.IGM.SizeTy, 2);
+  phi->addIncoming(llvm::ConstantInt::get(IGF.IGM.SizeTy, 0), prev);
+
+  // If we reach the end, jump to the continuation block.
+  auto *cond = IGF.Builder.CreateICmpULT(phi, dynamicLength);
+  IGF.Builder.CreateCondBr(cond, loop, rest);
+
+  IGF.Builder.emitBlock(loop);
+
+  auto *element =
+      emitPackExpansionElementWitnessTable(IGF, expansionTy, conformance, phi);
+
+  // Store the element metadata into to the current destination index.
+  auto *eltIndex = IGF.Builder.CreateAdd(dynamicIndex, phi);
+  Address eltPtr(IGF.Builder.CreateInBoundsGEP(pack.getElementType(),
+                                               pack.getAddress(), eltIndex),
+                 pack.getElementType(), pack.getAlignment());
+
+  IGF.Builder.CreateStore(element, eltPtr);
+
+  // Increment our counter.
+  auto *next =
+      IGF.Builder.CreateAdd(phi, llvm::ConstantInt::get(IGF.IGM.SizeTy, 1));
+
+  phi->addIncoming(next, loop);
+
+  // Repeat the loop.
+  IGF.Builder.CreateBr(check);
+
+  // Fall through.
+  IGF.Builder.emitBlock(rest);
+}
+
+StackAddress irgen::emitWitnessTablePack(IRGenFunction &IGF,
+                                         CanPackType packType,
+                                         PackConformance *packConformance) {
+  auto *shape = IGF.emitPackShapeExpression(packType);
+
+  if (auto *constantInt = dyn_cast<llvm::ConstantInt>(shape)) {
+    assert(packType->getNumElements() == constantInt->getValue());
+    return StackAddress(
+        emitFixedSizeWitnessTablePack(IGF, packType, packConformance));
+  }
+
+  assert(packType->containsPackExpansionType());
+  auto pack = IGF.emitDynamicAlloca(IGF.IGM.WitnessTablePtrTy, shape,
+                                    IGF.IGM.getPointerAlignment(),
+                                    /*allowTaskAlloc=*/true);
```

Comment:
I'm not sure I see where we undo this allocation.  We can't just stack-allocate and then assume the memory is valid forever: for one, we might need be in a loop or something similar where this could lead to stack exhaustion, and for another, there are other uses of dynamic stack allocation, and they do include deallocation, which will invalidate any later stack allocations we've made.  So anytime we do a dynamic stack allocation, we need to be set up to do a paired deallocation.  (And that is also the right place to add a lifetime.end intrinsic for the constant-size case.)

Reply:
And we have the same problem with type metadata at the moment.  Although `cleanupTypeMetadataPack` exists, it isn't called to clean up the dynamic allocation generated here
```
sil @h : $<T...> () -> () {
    %i = function_ref @i : $@convention(thin) <T...> () -> ()
    apply %i<Pack{repeat each T, repeat each T}>() : $@convention(thin) <T...> () -> ()
    %ret = tuple ()
    return %ret : $()
}
sil @i : $<T...> () -> () {
}
```
It seems likely we'd want to clean up both the dynamic allocation for type metadata stacks and witness table stacks at about the same time.

Reply:
I think the long-term solution here is probably:
- make the basic fetch operations in IRGen eagerly unique metadata/wtable packs on the heap unless we know we're in a context that doesn't need to
- Add a SIL pass to add explicit temporary allocations for metadata/wtable packs, and then preferentially use those in IRGen instead of eagerly uniquing.

Basically, the pass would figure out reasonable scopes in which to add `alloc_type_pack`, `init_type_pack`, and `dealloc_type_pack` instructions.  `alloc` and `dealloc` would just allocate an uninitialized array, and `init` would fill it in and set the pack as local type data, which would make dominated uses in IRGen naturally use it.  I'm not sure what the state of the art is on inserting alloc/dealloc scopes.

Reply:
Makes sense.  Would like to land this for now so we can finish the lowering for `open_pack_element` ( https://github.com/apple/swift/pull/63502 ) and the other instructions.  After that, the heap allocation for metadata/wtables can be done.a